File: services/backend/src/sysmind/domain/agent_tasks.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Literal, TypeAlias

AgentTaskStatus: TypeAlias = Literal[
    "created",
    "planning",
    "running_tools",
    "analyzing",
    "waiting_user_input",
    "completed",
    "cancelling",
    "cancelled",
    "failed",
    "timed_out",
    "interrupted",
]
AgentEventType: TypeAlias = Literal[
    "task.created",
    "task.status",
    "provider.started",
    "provider.completed",
    "tool.started",
    "tool.completed",
    "tool.failed",
    "task.message",
    "task.completed",
    "task.failed",
]


@dataclass(frozen=True, slots=True)
class AgentBudget:
    max_rounds: int = 4
    max_tool_calls: int = 8
    timeout_seconds: float = 30.0
    max_parallel_tools: int = 2

    def __post_init__(self) -> None:
        # Mirrors AgentTaskManager.start so an invalid budget fails at construction,
        # not halfway through wiring a task.
        if self.max_rounds < 1:
            raise ValueError("max_rounds must be at least 1.")
        if self.max_tool_calls < 1:
            raise ValueError("max_tool_calls must be at least 1.")
        if self.timeout_seconds <= 0:
            raise ValueError("timeout_seconds must be positive.")
        if self.max_parallel_tools < 1:
            raise ValueError("max_parallel_tools must be at least 1.")


@dataclass(frozen=True, slots=True)
class AgentTaskRecord:
    id: str
    status: AgentTaskStatus
    user_goal: str
    provider: str
    allowed_tools: tuple[str, ...]
    budget: AgentBudget
    current_round: int
    tool_call_count: int
    progress: int
    working_summary: dict[str, object]
    final_output: str | None
    failure_code: str | None
    failure_message: str | None
    cancel_requested: bool
    created_at: str
    started_at: str | None
    finished_at: str | None
    schema_version: str

    def __post_init__(self) -> None:
        object.__setattr__(self, "working_summary", dict(self.working_summary))


@dataclass(frozen=True, slots=True)
class AgentTaskEvent:
    id: int
    task_id: str
    event_type: AgentEventType
    data: dict[str, object]
    created_at: str

    def __post_init__(self) -> None:
        object.__setattr__(self, "data", dict(self.data))


@dataclass(frozen=True, slots=True)
class AgentToolCallRecord:
    id: str
    task_id: str
    provider_call_id: str
    tool_name: str
    tool_version: str
    status: str
    arguments_hash: str
    risk_level: str
    started_at: str
    finished_at: str | None
    duration_ms: int | None
    result_summary: dict[str, object] | None
    error_code: str | None
    error_message: str | None

    def __post_init__(self) -> None:
        if self.result_summary is not None:
            object.__setattr__(self, "result_summary", dict(self.result_summary))

File: apps/desktop/src/services/agent-tasks.ts
import type { ApiClient, SseEvent } from "./api-client";

const MAX_RECONNECT_DELAY_MS = 5_000;

export function agentTaskReconnectDelay(failedAttempts: number): number {
  return Math.min(350 * 2 ** Math.max(0, failedAttempts - 1), MAX_RECONNECT_DELAY_MS);
}

export type AgentTaskStatus =
  | "created"
  | "planning"
  | "running_tools"
  | "analyzing"
  | "waiting_user_input"
  | "completed"
  | "cancelling"
  | "cancelled"
  | "failed"
  | "timed_out"
  | "interrupted";

export interface ToolDescriptor {
  name: string;
  version: string;
  qualified_name: string;
  description: string;
  input_schema: Record<string, unknown>;
  risk_level: string;
  sensitivity: string[];
}

export interface AgentTask {
  id: string;
  status: AgentTaskStatus;
  user_goal: string;
  provider: string;
  allowed_tools: string[];
  budget: {
    max_rounds: number;
    max_tool_calls: number;
    timeout_seconds: number;
    max_parallel_tools: number;
  };
  current_round: number;
  tool_call_count: number;
  progress: number;
  final_output: string | null;
  failure_code: string | null;
  failure_message: string | null;
  cancel_requested: boolean;
  created_at: string;
  started_at: string | null;
  finished_at: string | null;
  schema_version: string;
  tool_calls: Array<{
    id: string;
    tool_name: string;
    tool_version: string;
    status: string;
    duration_ms: number | null;
    result_summary: Record<string, unknown> | null;
    error_code: string | null;
    error_message: string | null;
  }>;
}

export interface AgentTaskEventData {
  status?: AgentTaskStatus;
  progress?: number;
  tool?: string;
  action?: string;
  message?: string;
  error_code?: string;
  created_at: string;
}

export function getToolCatalog(
  client: ApiClient,
  signal?: AbortSignal,
): Promise<{ items: ToolDescriptor[] }> {
  return client.get<{ items: ToolDescriptor[] }>("/api/v1/tasks/tools", signal);
}

export function startAgentTask(
  client: ApiClient,
  userGoal: string,
  allowedTools: string[],
  signal?: AbortSignal,
): Promise<AgentTask> {
  return client.postJson<AgentTask, Record<string, unknown>>(
    "/api/v1/tasks",
    { user_goal: userGoal, allowed_tools: allowedTools },
    signal,
  );
}

export function getAgentTask(
  client: ApiClient,
  taskId: string,
  signal?: AbortSignal,
): Promise<AgentTask> {
  return client.get<AgentTask>(`/api/v1/tasks/${taskId}`, signal);
}

export function getRecentAgentTasks(
  client: ApiClient,
  signal?: AbortSignal,
): Promise<{ items: AgentTask[] }> {
  return client.get<{ items: AgentTask[] }>("/api/v1/tasks", signal);
}

export function cancelAgentTask(
  client: ApiClient,
  taskId: string,
  signal?: AbortSignal,
): Promise<AgentTask> {
  return client.post<AgentTask>(`/api/v1/tasks/${taskId}/cancel`, signal);
}

export function streamAgentTaskEvents(
  client: ApiClient,
  taskId: string,
  onEvent: (event: SseEvent<AgentTaskEventData>) => void,
  signal?: AbortSignal,
  lastEventId?: string,
): Promise<void> {
  return client.streamSse(
    `/api/v1/tasks/${taskId}/events`,
    onEvent,
    signal,
    lastEventId,
  );
}
